crates/goose/src/truncate.rs

```diff
@@ -0,0 +1,706 @@
+    fn determine_indices_to_remove(
+        &self,
+        messages: &[Message],
+        token_counts: &[usize],
+        context_limit: usize,
+    ) -> Result<HashSet<usize>> {
+        let mut indices_to_remove = HashSet::new();
+        let mut total_tokens: usize = token_counts.iter().sum();
+        let mut tool_ids_to_remove = HashSet::new();
+
+        for (i, message) in messages.iter().enumerate() {
+            if total_tokens <= context_limit {
+                break;
+            }
+
+            // Remove the message
+            indices_to_remove.insert(i);
+            total_tokens -= token_counts[i];
+            debug!(
+                "OldestFirst: Removing message at index {}. Tokens removed: {}",
+                i, token_counts[i]
+            );
+
+            // If it's a ToolRequest or ToolResponse, mark its pair for removal
+            if message.is_tool_call() || message.is_tool_response() {
+                if let Some(tool_id) = message.get_tool_id() {
+                    tool_ids_to_remove.insert((i, tool_id.to_string()));
+                }
+            }
+        }
+
+        // Now, find and remove paired ToolResponses or ToolRequests
+        for (i, message) in messages.iter().enumerate() {
+            if let Some(tool_id) = message.get_tool_id() {
+                // the other part of the pair has same tool_id but different message index
+                if tool_ids_to_remove
+                    .iter()
+                    .any(|&(idx, ref id)| id == tool_id && idx != i)
+                {
+                    indices_to_remove.insert(i);
+                    total_tokens -= token_counts[i];
+                }
+            }
+        }
+
+        Ok(indices_to_remove)
+    }
+}
+
+/// Strategy to truncate messages by removing from the middle outwards
+pub struct MiddleOutTruncation;
+
+impl TruncationStrategy for MiddleOutTruncation {
+    fn determine_indices_to_remove(
+        &self,
+        messages: &[Message],
+        token_counts: &[usize],
+        context_limit: usize,
+    ) -> Result<HashSet<usize>> {
+        let mut total_tokens: usize = token_counts.iter().sum();
+        let mut left = messages.len() / 4; // Start at ~1/4 point
+        let mut right = messages.len() * 3 / 4; // End at ~3/4 point
+
+        // Much easier to have user messages on the left
+        // and assistant messages on the right
+        if left > 0 && messages[left].role == Role::Assistant {
+            left -= 1; // guaranteed to be user since we alternate
+        }
+
+        if right < messages.len() - 1 && messages[right].role == Role::User {
+            right += 1; // guaranteed to be assistant
+        }
+
+        // If after removing tokens between 1/4 and 3/4, we would still exceed the context limit,
+        // Then its better to just use the oldest first strategy (recent messages are more important)
+        let middle_half_tokens = token_counts[left..=right].iter().sum::<usize>();
```

Comment:
similar comment as above, maybe we don't even check this and let it recurse if necessary?